### PR TITLE
New Workflow: `cp_process_singlecells` !

### PR DESCRIPTION
**Pull request by axiomcura**
## About

This PR introduces a new a workflow that is based on @jenna-tomkinson single cell feature extraction [analysis](https://github.com/WayScience/NF1_SchwannCell_data/blob/7b0f5a232ddaf87a92d93bc3a7d23be666cd3627/4_processing_features/extraction_utils.py#L9)

### What’s new

- New module `cytotable_convert.smk`
    - Converts `sqlite` files into `parquet` files. For more information about `CytoTable`, please look at the [repo](https://github.com/cytomining/CytoTable)
    - Workflow that contains 3 steps.
        - First step converts `sqlite` files into `paruquet` files by using `cytostable_convert.smk` module
        - Next is to normalize the data set by using `pycytominer`'s [normalization](https://github.com/cytomining/pycytominer/blob/master/pycytominer/normalize.py) function within the `normalize.smk` module
        - Then features are extracted using `pycytominer`'s [feature extraction](https://github.com/cytomining/pycytominer/blob/master/pycytominer/feature_select.py) function in the `feature_selection.smk` module
- Introducing to workflow config file
    - Workflow configs contains all the parameters used within the workflow
    - This is located in `configs/wl_configs/cp_process_singlecells.yaml`
        - NOTE: workflow configs names will have the same name as the workflow

## Usage

To use this workflow first one must initalize the data set for 

```
cytosnake init -d *.sqlite -m metadata -b barcodes.txt
```

NOTE: Make sure to replace the file names with your own file names. 

Where:

- `-d` is the dataset, it can also be a list of datasets
- `-m` refers to the metadata folder
- `-b` refers to the barcodes files, however, this is optional. If your datasets do not come with barcodes, then this will be defaulted to `None`

This will make `CytoSnake` recognize that in the current directory you are in is the `ProjectDirectory` 

once the initialization step is complete, then you can use `CytoSnake`'s `run` mode to execute the `cp_process_singlecells` workflow 

```
cytosnake run cp_process_singlecells
```

In your `ProjecDirectory`, a new directory will appear titled as `results`. That is where all the outputs will be saved. 

- NOTE: the converted datasets will be in the `data` folder

And that’s it!

### Change configurations.

To change the `cp_process_singlecells` configs, go to `configs/wl_configs/cp_process_singlecells.yaml` file 

The structure of the config is:

```yaml
name: name_of_workflow

module1_configs:
  params:
    - convert: parquet
module2_configs:
  params:
    - method: spherize 
# and so on
```

To change the parameters, change the values to each keyword under the `params` section. Make sure the inputs are valid. Please refer to the software's documentation.. `CytoSnake` cannot preemptively check if the new parameters are valid before executing the workflow.

### CODE REVIEW

File: workflows/scripts/normalize.py
Comment (d33bs) [inline workflows/scripts/normalize.py:55]: Consider removing these comments or adding it into more formal documentation somehow.

File: cytosnake/utils/cyto_paths.py
Comment (d33bs) [inline cytosnake/utils/cyto_paths.py:164]: Could the code specified in the return be used more directly to avoid creating a new function? Doing this may simplify the code and reduce the complexity.

File: configs/configuration.yaml
Comment (d33bs) [inline configs/configuration.yaml:16]: What do general configurations provide in context with the other configurations below? Consider providing a description of this somewhere.

File: configs/wf_configs/cp_process_singlecells.yaml
Comment (d33bs) [inline configs/wf_configs/cp_process_singlecells.yaml:1]: Really neat to see these configuration elements in one spot! Consider providing some documentation within this file to distinguish what the overall purpose is and perhaps what each chunk applies to.

File: workflows/scripts/feature_select.py
Comment (d33bs) [inline workflows/scripts/feature_select.py:75]: Noticing the difference here from `workflows/scripts/convert.py`, consider keeping the approach with files under `workflows/scripts/...` consistent with one another to help reduce the overall complexity of maintaining these.

**Review by d33bs — COMMENTED** [review]
Nice job! I left a few comments and suggestions on the code found in this PR. I plan to follow up with more functional feedback based on the `Usage` guide provided in the PR description. Overall I thought things looked good and began to wonder about the following items:

- Would it make sense to move some of the more formal content under the PR description to documentation for the project?
- How would you feel about adding a test for this code to help build confidence and document the capabilities involved with code? This could possibly follow the example provided from the PR description under `Usage`.

I'm interested in discussing these and other elements from the review, please don't hesitate to let me know if you have any questions along the way.

File: configs/analysis_configs/cytotable_convert.yaml
Comment (gwaybio) [inline configs/analysis_configs/cytotable_convert.yaml:6]: This is probably beyond scope of `cytosnake`, but can you specify non-canonical compartments? (e.g. anything beyond Cells, Cytoplasm, and Nuclei)

File: configs/configuration.yaml
Comment (gwaybio) [reply configs/configuration.yaml:16]: yes, I agree. I think being more specific about this file name (`configuration.yaml`) might help to clarify as well.

File: cytosnake/guards/ext_guards.py
Comment (gwaybio) [inline cytosnake/guards/ext_guards.py:2]: I'm a bit confused by the strategy/philosophy of this file. Why does `CytoSnake` need to guard against this at all? Isn't our philosophy to have the tools that `CytoSnake` uses handle this? (e.g. pycytominer cytomining/pycytominer#262)

File: cytosnake/guards/path_guards.py
Comment (gwaybio) [inline cytosnake/guards/path_guards.py:12]: I agree that `CytoSnake` needs to guard paths! (in reference to an earlier comment about `CytoSnake` guarding file types)

File: cytosnake/helpers/helper_funcs.py
Comment (gwaybio) [inline cytosnake/helpers/helper_funcs.py:205]: it seems to me that the only purpose of this file is to append a hardcoded `.parquet` to `output_name`.

This to me smells too complex and you might benefit (in regards to maintainability) from thinking through a simpler solution

File: cytosnake/utils/config_utils.py
Comment (gwaybio) [inline cytosnake/utils/config_utils.py:50]: if you change the name of the this file above, also make sure to propagate the change here.

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [reply cytosnake/utils/cyto_paths.py:164]: agree that we want to reduce complexity! This is already quite complex!

I understand snakemake is particular about things, and this might be one of them, so please feel free to decide how to proceed.

File: workflows/rules/common.smk
Comment (gwaybio) [inline workflows/rules/common.smk:29]: `CONVERTED` is not specific. Can you think of a more explicit name? maybe `CYTOTABLE_OUTPUT_DATA`?

File: workflows/scripts/convert.py
Comment (gwaybio) [inline workflows/scripts/convert.py:20]: it doesn't always. I think CytoTable can handle CSVs distributed across multiple folders. Please correct me if i'm wrong @d33bs !

File: workflows/scripts/convert.py
Comment (gwaybio) [reply workflows/scripts/convert.py:41]: i agree - unless it is snakemake specific, please try to reduce complexity - this will help SO much in the future (avoiding headaches with debugging, updating for new functionality, general gardening, etc.)

File: workflows/scripts/feature_select.py
Comment (gwaybio) [inline workflows/scripts/feature_select.py:45]: this is a great example of reducing complexity!! Simplicity is key!! :)

**Review by gwaybio — CHANGES_REQUESTED** [review]
Looking great! 🤩 

My general comment is to please reduce complexity. You may consider doing this here (probably in some cases this is within scope) but you (and the lab!) will thank yourself later for making this project easier to read and maintain.

I also agree that the documentation in this PR is great!! (as it has been for many of the past PRs) I do hope that you are porting this documentation elsewhere more forward facing. We don't want to forget about this content!

File: workflows/scripts/normalize.py
Comment (jenna-tomkinson) [inline workflows/scripts/normalize.py:46]: Just wondering, where did you get these metadata column names from? Based on my experience, the metadata columns will always vary by project so I assume that somewhere if you need to access metadata that this is made more flexible?

File: workflows/scripts/normalize.py
Comment (jenna-tomkinson) [inline workflows/scripts/normalize.py:63]: Are meta_features specified in the config file?

Is there no way to have it automatically find columns in a file starting with Metadata (which CellProfiler consistently names all metadata columns this way) instead of a user having to name all of the metadata columns?

Also, the name `meta_features` is a bit confusing to me. From what I can see from the commented-out code, it is just the metadata column names so I would recommend to make the name clearer.

**Review by jenna-tomkinson — COMMENTED** [review]
Since both Greg and Dave got into the nitty gritty of the code, I tried to find things that didn't make sense to me, so I only have a few comments about that.

Other than that, I am VERY excited to get to test this! 😄

File: configs/analysis_configs/cytotable_convert.yaml
Comment (axiomcura) [reply configs/analysis_configs/cytotable_convert.yaml:4]: Ah yes, It definitely does! 

The way how these configs files are designed is that we take the default parameters of a function and place them into the config file. 

Looking at [CytoTable’s](https://github.com/cytomining/CytoTable/blob/main/cytotable/convert.py) `convert` workflow, it seems that `source_datatype` is set to `None`. However, you are correct. It is more practical for the configs to have their own defaults. Like in this case, making `source_datatype` as `sqlite` instead of `null` since the expected inputs are `sqlite` files.

Thanks for pointing this out!

File: configs/analysis_configs/cytotable_convert.yaml
Comment (axiomcura) [reply configs/analysis_configs/cytotable_convert.yaml:6]: This will depend on `CytoTable`'s capability. The config files will take anything as long as the parameters are valid in `CytoTable`'s `convert` workflow. Therefore, if there is support for non-canonical compartments within `CytoTable` then it should work? 

Looking at the documentation, the `join` [parameter](https://github.com/cytomining/CytoTable/blob/f336d26baf96b8b3d9ed6ca07f617895030bf038/cytotable/convert.py#L861-L862) only takes a boolean. 

Correct me if I am wrong @d33bs.This will depend on `CytoTable`'s capability. The config files will take anything as long as the parameters are valid in `CytoTable`'s `convert` workfow. Therefore, if there is support for non-cononical compartments within `CytoTable` then it should work? 

Looking at the documentation, the `join` [parameter](https://github.com/cytomining/CytoTable/blob/f336d26baf96b8b3d9ed6ca07f617895030bf038/cytotable/convert.py#L861-L862) only takes a boolean. 

Correct me if I am wrong @d33bs.

File: configs/configuration.yaml
Comment (axiomcura) [reply configs/configuration.yaml:16]: I have been thinking of changing the name as well. I will write an issue pertaining to this because I do not know if changing the name of the `configurational` file will cause a domino effect of pathing errors. 

To provide some context. The `configurational` contains the settings for only the `CytoSnake` CLI. This will include presets used to execute a workflow, number of cores, stating data format, etc. Anything that has to do with analysis, those are found within the workflow configs `wf_configs`. 

I am thinking of changing the name to `cytosnake_configs.yaml`

Comment (axiomcura) [reply configs/configuration.yaml:16]: Issue created #38

Comment (axiomcura) [reply configs/configuration.yaml:16]: I have added documentation at the top of the workflow. Let me know if this is a suitable method to display the workflow documentation.

File: configs/wf_configs/cp_process_singlecells.yaml
Comment (axiomcura) [reply configs/wf_configs/cp_process_singlecells.yaml:1]: added documentation!

File: cytosnake/guards/ext_guards.py
Comment (axiomcura) [reply cytosnake/guards/ext_guards.py:2]: It's true that `pycytominer` has its internal checking systems for file extensions. However, this module is mainly used for `CytoSnake` pathing system and potentially other software packages that may not check for format extensions. 

Could be that this module will removed in the future.

Also, I may have written this module before (pycytominer cytomining/pycytominer#262) I do not remember.

File: cytosnake/guards/ext_guards.py
Comment (axiomcura) [reply cytosnake/guards/ext_guards.py:13]: Updated!

File: cytosnake/guards/ext_guards.py
Comment (axiomcura) [reply cytosnake/guards/ext_guards.py:33]: Updated!

File: cytosnake/guards/path_guards.py
Comment (axiomcura) [reply cytosnake/guards/path_guards.py:24]: Correct! thanks for catching that!

File: cytosnake/guards/path_guards.py
Comment (axiomcura) [reply cytosnake/guards/path_guards.py:12]: Ah okay! Thanks!

File: cytosnake/helpers/helper_funcs.py
Comment (axiomcura) [reply cytosnake/helpers/helper_funcs.py:205]: Yeah I've noticed that too as well. This is mentioned in [#32](https://github.com/WayScience/CytoSnake/issues/32)

I will be creating a seperate PR that will fix this issue.

File: cytosnake/helpers/helper_funcs.py
Comment (axiomcura) [reply cytosnake/helpers/helper_funcs.py:215]: Yeah, now looking in retrospect, the helper functions are not well developed as it takes in too much repeated code. 

Will be creating a PR for this issue: #32

File: cytosnake/helpers/helper_funcs.py
Comment (axiomcura) [reply cytosnake/helpers/helper_funcs.py:219]: This will be updated in the future! [#32](https://github.com/WayScience/CytoSnake/issues/32)

File: cytosnake/utils/config_utils.py
Comment (axiomcura) [reply cytosnake/utils/config_utils.py:50]: Correct. Any name changes that occurs will also have to be reflected here. Still thinking on ideas on how to prevent harcoding file names. I am open to any suggestions

File: cytosnake/utils/config_utils.py
Comment (axiomcura) [reply cytosnake/utils/config_utils.py:51]: Removed!

File: cytosnake/utils/cyto_paths.py
Comment (axiomcura) [reply cytosnake/utils/cyto_paths.py:110]: Good point! Thanks!

File: workflows/scripts/convert.py
Comment (axiomcura) [reply workflows/scripts/convert.py:17]: added!

File: workflows/scripts/convert.py
Comment (axiomcura) [reply workflows/scripts/convert.py:41]: I agree, this is unnecessarily complex for just a single function call.

I've been thinking about this quite some time. Since the inputs are ["magically" imported](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#python) into the script, I feel that the main script should handle everything. 

I think this specific design will only work if we have multiple functions. As of right now, only one function is being used. 

I'll make this as an issue. #41

File: workflows/scripts/convert.py
Comment (axiomcura) [reply workflows/scripts/convert.py:48]: Sounds good! here's also [documentation](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#python) on how `snakemake` gets imported into scripts.

File: workflows/scripts/feature_select.py
Comment (axiomcura) [reply workflows/scripts/feature_select.py:26]: added!

File: workflows/scripts/feature_select.py
Comment (axiomcura) [reply workflows/scripts/feature_select.py:75]: Definitely, I wrote an issue in regards of the inconsistency. [#40](https://github.com/WayScience/CytoSnake/issues/40)

File: workflows/scripts/normalize.py
Comment (axiomcura) [reply workflows/scripts/normalize.py:46]: This was obtained from the [cell-health study](https://github.com/broadinstitute/cell-health/blob/30ea5de393eb9cfc10b575582aa9f0f857b44c59/1.generate-profiles/generate_profiles.py#L62-L71) that used hard coded feature names. I think this will no longer be needed anymore since we can write this down in the config file.

File: workflows/scripts/normalize.py
Comment (axiomcura) [reply workflows/scripts/normalize.py:55]: Removed! I think this can just be added into the config file if needed

File: workflows/scripts/normalize.py
Comment (axiomcura) [reply workflows/scripts/normalize.py:63]: Yes meta_features is defined in the workflow config file. To answer your question about the name, we are keeping the name consistent with the function parameters. This will make it easier for users to search it up in the documentation. 

Here's the [documentation](https://github.com/cytomining/pycytominer/blob/a5ae6c81a275b692ef5d4c85cfeb37696bf69242/pycytominer/normalize.py#L44-L47) if you're curious. 

It seems that `pycytominer` get's it's meta features from column names. I do not know if it can do what you're asking. Maybe @gwaybio may have a better answer. In addition, you can overwrite the `infer` default parameter in `meta_features` and provide your own list of features in the config.

**Comment by axiomcura** [comment]
Hello @axiomcura @jenna-tomkinson @d33bs 

Hopefully I have attended all your comments and suggestions. Ready for another review!

File: cytosnake/guards/ext_guards.py
Comment (gwaybio) [reply cytosnake/guards/ext_guards.py:2]: Sounds good! Thanks for thinking this through! Will be good to keep in mind for future

File: cytosnake/utils/config_utils.py
Comment (gwaybio) [reply cytosnake/utils/config_utils.py:50]: i think it is ok to hardcode some things - ultimately, they will need to be hardcoded somewhere!

You may consider having a single function where these names are hardcoded, and calling the function wherever hardcodes are used. Just don't update the corresponding dictionary keys!

I think what you have now is fine.

File: workflows/scripts/normalize.py
Comment (gwaybio) [reply workflows/scripts/normalize.py:51]: pycytominer identifies `meta_features` using `Metadata_` column name prefixes if and only if `meta_features='infer'`. 

We will almost always use `meta_features='infer'`, but we want flexibility to define metadata columns if necessary

File: cytosnake/utils/cyto_paths.py
Comment (gwaybio) [reply cytosnake/utils/cyto_paths.py:164]: Hacky is ok sometimes!

**Review by gwaybio — APPROVED** [review]
Looks like this PR spawned several other issues!

Of the ones it spawned, which are highest priority in terms of handing this off to @jenna-tomkinson to play with?

**Comment by axiomcura** [comment]
> Looks like this PR spawned several other issues!
> 
> Of the ones it spawned, which are highest priority in terms of handing this off to @jenna-tomkinson to play with?

It's too hard to tell. My best bet is which ever @jenna-tomkinson finds the most annoying. 😅

**Comment by gwaybio** [comment]
oh gotcha! So you're saying after this is merged, she can have at it? (i.e. no other issues need addressing?)

**Comment by axiomcura** [comment]
I have been using the `NF1` dataset to test `cp_process_singlecells` and it has been working for me. Hopefully everything goes smoothly for Jenna with the other datasets. hehe 😅 

I think one issue would be trying to run `CytoSnake` with multiple `metadata` directories in one single run

**Comment by axiomcura** [comment]
I think the plan right now is:

I send it to @jenna-tomkinson , she writes some issues in the issues section in the Repo and I attend those a first priority issues to take care off. 


>Glad to hear it!! I wonder if you can contribute this directly then... what do you think?
Don't mind at all. @jenna-tomkinson and I can sit together, tinker with the config files to generate high quality features :)!

File: workflows/scripts/convert.py
Comment (d33bs) [reply workflows/scripts/convert.py]: What other formats might you want to use with CytoSnake? If CytoSnake will be tightly coupled with the SQLite format, consider providing those details upfront via documentation.

File: workflows/scripts/convert.py
Comment (d33bs) [reply workflows/scripts/convert.py:72]: Given Snakemake's specificity and how it differs from other workflow implementations, it may be worth linking to that documentation within the code via a comment or within this project's documentation somehow. I respect your best judgement on this.

File: workflows/scripts/convert.py
Comment (d33bs) [reply workflows/scripts/convert.py:65]: @axiomcura - Thanks for addressing this, for now it might be helpful to link to the `magically imported` input documentation you mentioned. This could help rationalize future development decisions if they are needed before #41 is resolved.

File: workflows/scripts/convert.py
Comment (d33bs) [reply workflows/scripts/convert.py:84]: Seeing `__main__` in this way within Python files sometimes indicates there was a reason things were designed in that way. For example, [Dask documentation](https://docs.dask.org/en/stable/scheduling.html?highlight=__main__#dask-distributed-local) makes a special note about the design patterns surrounding the use of `__main__`.

Here, a single line comment justifying why it was used here could help with development decision making in the future. If it's not needed, what @gwaybio mentions about the added steps may warrant simplification.

File: workflows/scripts/convert.py
Comment (d33bs) [reply workflows/scripts/convert.py:44]: @axiomcura , @gwaybio - Yes, this is correct. CytoTable will attempt to navigate a provided source data directory for nested CSV's or SQLite data and concatenate or join the data together in the results.

File: configs/analysis_configs/cytotable_convert.yaml
Comment (d33bs) [inline configs/analysis_configs/cytotable_convert.yaml:4]: Double checking: does this mean CytoSnake is tightly coupled to SQLite source data input for CytoTable?

File: configs/configuration.yaml
Comment (d33bs) [inline configs/configuration.yaml:1]: Should this have a new name provided?

File: configs/wf_configs/cp_process_singlecells.yaml
Comment (d33bs) [inline configs/wf_configs/cp_process_singlecells.yaml:4]: 🥳 Love the dedication to audience understanding here!

File: configs/wf_configs/cp_process_singlecells.yaml
Comment (d33bs) [inline configs/wf_configs/cp_process_singlecells.yaml:18]: Consider referencing CytoTable's docsite API reference here: https://cytomining.github.io/CytoTable/python-api.html#cytotable.convert.convert . The difference is minor, but may increase the possibility for understanding.

File: cytosnake/guards/path_guards.py
Comment (d33bs) [inline cytosnake/guards/path_guards.py:24]: 🙌

File: workflows/rules/common.smk
Comment (d33bs) [inline workflows/rules/common.smk:30]:
```suggestion
# Contains file path of the parquet datasets that are obtained from CytoTable
```

File: workflows/rules/common.smk
Comment (d33bs) [inline workflows/rules/common.smk:32]: The specificity here is great!

File: workflows/scripts/convert.py
Comment (d33bs) [inline workflows/scripts/convert.py:33]: Minor comment: consider adjusting the newlines here (depending on what's aligned with linting).

File: workflows/scripts/feature_select.py
Comment (d33bs) [inline workflows/scripts/feature_select.py:26]: 👍

**Review by d33bs — APPROVED** [review]
Nice work! Thank you for addressing the comments and for making many changes in addition to the new issues! I left a few minor comments + responses and overall felt this LGTM, respecting your decision on when it's best to merge.

File: workflows/scripts/convert.py
Comment (axiomcura) [reply workflows/scripts/convert.py]: I did change the name to a more general term `convert_to_parquet`. But you are right, currently the `CytoTable` module in `CytoSnake` only takes `sqlite` files as input. 

I've looked at `CytoTable`'s documentation and it seems that there other supported file formats that can be converted into `parquet` files. 

I will create an issue in regards to this. #43  

Thanks!

File: workflows/rules/common.smk
Comment (axiomcura) [reply workflows/rules/common.smk]: Documentation added!

File: workflows/rules/common.smk
Comment (axiomcura) [reply workflows/rules/common.smk]: Updated! I've also added documentation about the data level these variable names belong to.

File: workflows/scripts/convert.py
Comment (axiomcura) [reply workflows/scripts/convert.py:44]: Sound good! created issue for future implementation #43

File: configs/configuration.yaml
Comment (axiomcura) [reply configs/configuration.yaml:1]: updated!

**Comment by axiomcura** [comment]
@d33bs @gwaybio Hopefully I have attended all comments and suggestions. If not, please let me know! 

Thank you for your inputs! made some issues that I will take care in the future!

File: cytosnake/utils/config_utils.py
Comment (axiomcura) [reply cytosnake/utils/config_utils.py:50]: I was thinking of making a `class` the handles all the file pathing function calls. It's an idea.

**Pull request merged**
